File: algorithm/[book]c_algorithm/src/h.8-1_simple_hash_table/SimpleHashTable.c
/*
 * SimpleHashTable.c
 *
 *  Created on: 2019. 1. 7.
 *      Author: otrodevym
 */

#include "SimpleHashTable.h"

HashTable* SHT_CreateHashTable(int TableSize){
	HashTable* HT = (HashTable*) malloc(sizeof(HashTable));
	HT->Table = (Node*)malloc(sizeof(Node)*TableSize);
	HT->TableSize = TableSize;

	return HT;
}
void SHT_Set(HashTable* HT, KeyType Key, ValueType Value){
	int Address = SHT_Hash(Key, HT->TableSize);
	HT->Table[Address].Key = Key;
	HT->Table[Address].Value = Value;
}
ValueType SHT_Get(HashTable* HT, KeyType Key){
	int Address = SHT_Hash(Key, HT->TableSize);
	return HT->Table[Address].Value;
}
void SHT_DestroyHashTable(HashTable* HT){
	free(HT->Table);
	free(HT);
}
int SHT_Hash(int Key, int TableSize){
	return Key % TableSize;
}

int MyHash(char* ChKey, int KeyLength, int TableSize){
	int i = 0;
	int HashValue = 0;

	for(i = 0; i < KeyLength; i++){
		HashValue += (HashValue<<3) + ChKey[i];
	}
	return HashValue % TableSize;
}

File: algorithm/Java_Algorithm/src/h_hash_table/SimpleHashTable.java
package h_hash_table;


public class SimpleHashTable {
	private HashNode[] table;
	private int size;
	
	public SimpleHashTable() {
		
	}
	public SimpleHashTable(int size) {
		this.table = new HashNode[size];
		this.size = size;
	}
	private void shtSet(int key, int value) {
		int address = this.shtHash(key);
		this.table[address] = new HashNode();
		this.table[address].setKey(key);
		this.table[address].setValue(value);
	}
	
	private int shtGet(int key) {
		int address = this.shtHash(key);
		return this.table[address].getValue();
	}
	private int shtHash(int key) {
		return key % this.size;
	}
	
	
	public static void main(String args[]) {
		SimpleHashTable ht = new SimpleHashTable(193);
		
		ht.shtSet(418, 32114);
		ht.shtSet(9, 514);
		ht.shtSet(27, 8917);
		ht.shtSet(1031, 286);
		
		System.out.println("418 : " + ht.shtGet(418));
		System.out.println("9 : " + ht.shtGet(9));
		System.out.println("27 : " + ht.shtGet(27));
		System.out.println("1031 : " + ht.shtGet(1031));
		
		String[] str= new String[10];
		str[0] = "asdasd";
		str[1] = "qweqwe";
		str[2] = "zxczxc";
		
	}
	
}
